## Упражнения по библиотеке Numpy

In [1]:
import numpy as np
import time
from scipy.spatial.distance import cdist

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [ ]:
ar = np.random.randint(0,9,10)

In [ ]:
ar

array([0, 3, 6, 4, 0, 5, 2, 1, 4, 6])

In [ ]:
ar[(ar > 3) & (ar < 8)] *= -1

In [ ]:
ar

array([ 0,  3, -6, -4,  0, -5,  2,  1, -4, -6])

**2.** Заменить максимальный элемент случайного массива на 0

In [ ]:
ar[ar == ar.max()] = 0

In [ ]:
ar

array([ 0,  0, -6, -4,  0, -5,  2,  1, -4, -6])

**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [ ]:
matr = np.array([[1,2,3],[4,5,6]])
matr

array([[1, 2, 3],
       [4, 5, 6]])

In [ ]:
matr[0,:][None, :] * matr[1,:][:, None]

array([[ 4,  8, 12],
       [ 5, 10, 15],
       [ 6, 12, 18]])

**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [36]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 2, 9],
    [9, 9, 9],
    [1, 5, 7],
    [2, 2, 7],
    [4, 6, 7],
    [6, 4, 1]
])

B = np.array([
    [2, 7]
])

In [37]:
mask = np.all([np.any(A == elem, axis=1) for elem in B[0]], axis=0)

result = A[mask]
result

array([[7, 2, 9],
       [2, 2, 7]])

**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [13]:
matr = np.array([ [2,2,2],
                  [1,1,1],
                  [1,3,4],
                  [5,5,5],
                  [5,6,6],
                  [3,5,6],
                  [2,4,5],
                  [5,6,7],
                  [6,5,3],
                  [1,3,4] ])

In [15]:
matr[~np.all(matr == matr[:, [0]], axis=1)]


array([[1, 3, 4],
       [5, 6, 6],
       [3, 5, 6],
       [2, 4, 5],
       [5, 6, 7],
       [6, 5, 3],
       [1, 3, 4]])

**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [ ]:
matr = np.array([[1,2,3],
                 [4,5,6],
                 [1,2,3],
                 [8,9,0]])

In [ ]:
np.unique(matr, axis=0)

array([[1, 2, 3],
       [4, 5, 6],
       [8, 9, 0]])

______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [ ]:
x = [[1, 0, 1],
     [2, 0, 2],
     [3, 0, 3],
     [4, 4, 4]]

res = 1
for i in range(min(len(X), len(x[0]))):
    if x[i][i] != 0:
       res *= x[i][i]

res

3

In [ ]:
X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]])

X

array([[1, 0, 1],
       [2, 0, 2],
       [3, 0, 3],
       [4, 4, 4]])

In [ ]:
np.prod(np.diag(X)[np.diag(X) != 0])


np.int64(3)

* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [ ]:
 x = np.array([1, 2, 2, 4])
 y = np.array([4, 2, 1, 2])

In [ ]:
np.all(np.unique(x, return_counts=True)[1] == np.unique(y, return_counts=True)[1]) and \
np.all(np.unique(x, return_counts=True)[0] == np.unique(y, return_counts=True)[0])

np.True_

In [ ]:
 x = [1, 2, 2, 4]
 y = [4, 2, 1, 2]

In [ ]:
def value_counts(arr):
    hashT = {}
    for elem in arr:
        if elem in hashT:
            hashT[elem] += 1
        else:
            hashT[elem] = 1
    return hashT

value_counts(x) == value_counts(y)

True

* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль.
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [ ]:
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

max_v = float('-inf')

for i in range(1, len(x)):
    if x[i-1] == 0:
        max_v = max(max_v, x[i])
max_v

5

In [ ]:
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

In [ ]:
ind = (np.where(x == 0)[0]) + 1
ind = ind[ind < len(x)]
ind

array([3, 5, 6])

In [ ]:
np.max(x[ind])

np.int64(5)

* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [ ]:
x = np.array([2, 2, 2, 3, 3, 3, 5])

np.unique(x, return_counts=True)

(array([2, 3, 5]), array([3, 3, 1]))

In [ ]:
x = [2, 2, 2, 3, 3, 3, 5]

value_counts(x)

{2: 3, 3: 3, 5: 1}

* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [ ]:
X = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
Y = np.array([[9, 8, 7], [6, 5, 4], [3, 2, 1]])

In [ ]:
start_time = time.time()
np.sqrt(np.sum((X[:, np.newaxis, :] - Y) ** 2, axis=2))
time.time() - start_time

0.00018215179443359375

In [ ]:
np.sqrt(np.sum((X[:, np.newaxis, :] - Y) ** 2, axis=2))

array([[10.77032961,  5.91607978,  2.82842712],
       [ 5.91607978,  2.82842712,  5.91607978],
       [ 2.82842712,  5.91607978, 10.77032961]])

In [ ]:
start_time = time.time()
distances = np.sqrt(np.sum((X[:, np.newaxis, :] - Y) ** 2, axis=2))
time.time() - start_time

0.0001957416534423828

In [ ]:
distances

array([[10.77032961,  5.91607978,  2.82842712],
       [ 5.91607978,  2.82842712,  5.91607978],
       [ 2.82842712,  5.91607978, 10.77032961]])

In [ ]:
X = [[1, 2, 3], [4, 5, 6], [7, 8, 9]]
Y = [[9, 8, 7], [6, 5, 4], [3, 2, 1]]

In [ ]:
distances = []
start_time = time.time()
for x_row in X:
    row_distances = []
    for y_row in Y:
        distance = (sum((xi - yi) ** 2 for xi, yi in zip(x_row, y_row))) ** 0.5
        row_distances.append(distance)
    distances.append(row_distances)
time.time() - start_time

0.0001800060272216797

In [ ]:
distances

[[10.770329614269007, 5.916079783099616, 2.8284271247461903],
 [5.916079783099616, 2.8284271247461903, 5.916079783099616],
 [2.8284271247461903, 5.916079783099616, 10.770329614269007]]

_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [ ]:
calorie_stats = np.loadtxt("/content/sample_data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [ ]:
average_calories = np.mean(calorie_stats)
calories_difference = average_calories - 60

In [ ]:
calories_difference

np.float64(46.883116883116884)

3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [ ]:
calorie_stats_sorted = np.sort(calorie_stats)
calorie_stats_sorted

array([ 50.,  50.,  50.,  70.,  70.,  80.,  90.,  90.,  90.,  90.,  90.,
        90.,  90., 100., 100., 100., 100., 100., 100., 100., 100., 100.,
       100., 100., 100., 100., 100., 100., 100., 100., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 110., 110., 110., 110., 110., 110., 110.,
       110., 110., 110., 110., 120., 120., 120., 120., 120., 120., 120.,
       120., 120., 120., 130., 130., 140., 140., 140., 150., 150., 160.])

4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [ ]:
median_calories = np.median(calorie_stats)
median_calories

np.float64(110.0)

5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [ ]:
nth_percentile = np.percentile(calorie_stats, 3.3)
nth_percentile

np.float64(60.16)

6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [ ]:
more_calories = np.mean(calorie_stats > 60)
more_calories

np.float64(0.961038961038961)

7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [ ]:
calorie_std = np.std(calorie_stats)
calorie_std

np.float64(19.35718533390827)

8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Анализ показал, что большинство конкурентов (96,1%) содержат более 60 калорий на порцию, при медиане равной 110 калорий, так же менее 4% злаков на рынке столь же легкие, значит CrunchieMunchies это хороший выбор